seqr/views/apis/igv_api.py

```diff
@@ -1,14 +1,129 @@
+
+        matched_individuals = Individual.objects.filter(family__project=project, individual_id__in=individual_dataset_mapping.keys())
+        unmatched_individuals = set(individual_dataset_mapping.keys()) - {i.individual_id for i in matched_individuals}
+        if len(unmatched_individuals) > 0:
+            raise Exception('The following Individual IDs do not exist: {}'.format(", ".join(unmatched_individuals)))
+
+        info.append('Parsed {} rows in {} individuals from {}'.format(
+            sum([len(rows) for rows in individual_dataset_mapping.values()]), len(individual_dataset_mapping), filename))
+
+        existing_sample_files = defaultdict(set)
+        for sample in IgvSample.objects.select_related('individual').filter(individual__in=matched_individuals):
+            existing_sample_files[sample.individual.individual_id].add(sample.file_path)
+
+        unchanged_rows = set()
+        for individual_id, updates in individual_dataset_mapping.items():
+            unchanged_rows.update([
+                (individual_id, update['filePath']) for update in updates
+                if update['filePath'] in existing_sample_files[individual_id]
+            ])
+
+        if unchanged_rows:
+            info.append('No change detected for {} rows'.format(len(unchanged_rows)))
+
+        all_updates = []
+        for i in matched_individuals:
+            all_updates += [
+                dict(individualGuid=i.guid, **update) for update in individual_dataset_mapping[i.individual_id]
+                if (i.individual_id, update['filePath']) not in unchanged_rows
+            ]
+
+    except Exception as e:
+        return create_json_response({'errors': [str(e)]}, status=400)
+
+    response = {
+        'updates': all_updates,
+        'uploadedFileId': uploaded_file_id,
+        'errors': [],
+        'info': info,
+    }
+    return create_json_response(response)
+
+
+SAMPLE_TYPE_MAP = {
+    'bam': IgvSample.SAMPLE_TYPE_ALIGNMENT,
+    'cram': IgvSample.SAMPLE_TYPE_ALIGNMENT,
+    'bigWig': IgvSample.SAMPLE_TYPE_COVERAGE,
+    'junctions.bed.gz': IgvSample.SAMPLE_TYPE_JUNCTION,
+    'dcr.bed.gz': IgvSample.SAMPLE_TYPE_GCNV,
+}
+
+@pm_or_data_manager_required
+def update_individual_igv_sample(request, individual_guid):
+    individual = Individual.objects.get(guid=individual_guid)
+    project = individual.family.project
+    check_project_permissions(project, request.user, can_edit=True)
+
+    request_json = json.loads(request.body)
+
+    try:
+        file_path = request_json.get('filePath')
+        if not file_path:
+            raise ValueError('request must contain fields: filePath')
+
+        suffix = '.'.join(file_path.split('.')[1:])
+        sample_type = SAMPLE_TYPE_MAP.get(suffix)
+        if not sample_type:
+            raise Exception('Invalid file extension for "{}" - valid extensions are {}'.format(
+                file_path, ', '.join(SAMPLE_TYPE_MAP.keys())))
+        if not does_file_exist(file_path):
```

Comment:
changed this to directly check the file here (used to have a helper method but that didn't hep much)